synapse/rest/client/v2_alpha/keys.py

```diff
@@ -163,6 +163,18 @@ async def on_POST(self, request):
         device_id = requester.device_id
         timeout = parse_integer(request, "timeout", 10 * 1000)
         body = parse_json_object_from_request(request)
+
+        try:
+            device_keys = body["device_keys"]
+        except KeyError:
+            raise InvalidAPICallError("device_keys is a required body parameter")
+
+        for mxid, device_ids in device_keys.items():
```

Comment:
Should we first confirm that `device_keys` is a `dict`?

It would also be nice to add a check that items in the list are of the proper form.

We do something similar to this:

https://github.com/matrix-org/synapse/blob/ef4329f8daa25079beba0894a3e46015eab2cbd7/synapse/federation/federation_client.py#L1420-L1424

Reply:
I was erring on the side of not doing too much work here, in lieu of looking for a more systematic approach. Happy to do this though

Reply:
Also, in the quoted code sample we will happily accept the case when `via` is a string:

```python3
Python 3.9.6 (default, Jul 16 2021, 00:00:00) 
[GCC 11.1.1 20210531 (Red Hat 11.1.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = "hello, world"
>>> from collections import Sequence
<stdin>:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
>>> isinstance(x, Sequence)
True
>>> all(isinstance(c, str) for c in x)
True
>>> 
```